### PR TITLE
Update README to explain when not to use this library

### PR DESCRIPTION
**Pull request by alisdair**
By design, the `terraform-json` library is not always suitable for every application. Its name and prominence may cause those integrating with Terraform to assume otherwise.

This commit expands the README to attempt to clarify situations where the `terraform-json` library is not appropriate, explaining why and what to do instead.

### CODE REVIEW

**Review by apparentlymart — APPROVED** [review]
This seems like a good set of guidelines. :+1:

File: README.md [hunk 1]
```diff
@@ -15,7 +15,37 @@ This repository also serves as de facto documentation for the formats produced
...
+* **Writing JSON output**: the structures in this library are not guaranteed to emit
+  JSON data which is semantically equivalent to Terraform itself. If your application
+  must robustly write JSON data to be consumed by systems which expect Terraform's
+  format to be supported, you should implement your own custom encoders.
+* **Filtering or round-tripping**: the Terraform JSON formats are designed to be
+  forwards compatible, and permit new attributes to be added which may safely be
+  ignored by earlier versions of consumers. This library **drops unknown attributes**,
+  which means it is unsuitable for any application which intends to filter data
+  or read-modify-write data which will be consumed downstream. Any application doing
+  this will silently drop new data from new versions. For this applicaiton, you should
```
Comment (bendbennett) [inline README.md:40]:
```suggestion
  this will silently drop new data from new versions. For this application, you should
```

**Review by radeksimko — CHANGES_REQUESTED** [review]
Perhaps this is just a question of wording rather than misunderstanding of intentions but as one of the consumers we do expect some level of version compatibility I tried to explain in-line.

The remaining part looks reasonable to me.

I'm still not entirely sure that `terraform-json` and custom decoders have to be mutually exclusive choice though and that the language in the second paragraph has to be as strong as to _recommend against_ entirely, I would just expect some words of caution.

File: README.md [hunk 2]
```diff
@@ -15,7 +15,37 @@ This repository also serves as de facto documentation for the formats produced
...
+general purpose use.
+
+The Terraform core team **recommends against** using `terraform-json` if your
+application has any of the following requirements:
+
+* **Broad version compatibility**: each version this library represents a specific
+  snapshot of the [Terraform JSON output format](https://developer.hashicorp.com/terraform/internals/json-format),
+  and it often slightly lags Terraform itself. If you require compatibility with
+  multiple Terraform versions, we recommend implementing your own custom decoders
+  for the parts of the JSON format you need.
```
[the remark this answers is not in the document]
Comment (alisdair) [reply README.md]: Thanks! I've revised this paragraph to hopefully clarify what I meant.

**Comment by alisdair** [comment]
> I'm still not entirely sure that terraform-json and custom decoders have to be mutually exclusive choice though

Can you say more on this, and perhaps recommend a change to the wording?

> and that the language in the second paragraph has to be as strong as to recommend against entirely, I would just expect some words of caution.

The recommendation against using the library (certainly by default) is the Terraform core team's position. We've seen recent usage which has caused compatibility problems and would rather err on the side of avoiding its usage for cases where its value is ambiguous. This doesn't mean that `terraform-ls` or other users should migrate away.

As a reminder, the Terraform core team's philosophy about compatibility in general has been to publish and adhere to protocol specifications, rather than to supply and support library implementations. While we are now (jointly) maintaining this library, we want to be clear that this doesn't mean that we recommend its usage in all cases.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (apparentlymart) [reply README.md]: A crucial design problem for this library is that its "implementation" largely consists of declarative struct tags rather than actual code and so its capabilities are limited by what Go's JSON library implements itself.

This caution is based on real-world experience of folks doing the things enumerated here and having surprising (to them) problems when e.g. there was a new version of Terraform but they didn't upgrade the library. The intent here is only to try to make such situations _expected_ rather than surprising, so that anyone using this library has a better idea of what to expect and what maintenance burden they might be taking on by using it (i.e. the need to promptly upgrade if the format has changed in a significant way).

In the long run I'd like to see us take a fresh look at this library as something whose implementation is fully encapsulated (where the JSON decoding behavior is implementation details rather than  allowing callers to directly use `encoding/json`) and thus can make better guarantees about how things will evolve. But this documentation is intended as damage control in the meantime since naive use if this library has now been the cause of several maintenance problems elsewhere.

To your specific point, I think the way we've defined major and minor format revisions means that what you want should happen "for free": a subsequent minor version of the format should be backward compatible with its predecessors in the same major version and so this library will be able to interpret them as a result of the underlying contract, and not due to anything this library is or is not doing. It remains to be seen how this library could manage to evolve across a major format change (i.e. a 2.0 format) but thankfully we have no current plans to produce one and even if we did it would presumably be opt-in at the Terraform CLI level so that clients can still get a reasonable approximation of the latest 1.x format for some reasonable migration period anyway.

In other words, I don't think this documentation changes anything in ways that your comment is concerned about because that's a question of underlying format compatibility rather than this library's own compatibility. It will suffer backward compatibility problems only if Terraform CLI itself breaks its compatibility promises, which seems unlikely in the foreseeable future.

File: README.md [hunk 3]
```diff
@@ -15,7 +15,39 @@ This repository also serves as de facto documentation for the formats produced
...
+
+This library was built for a few specific applications, and is not intended for
+general purpose use.
+
+The Terraform core team **recommends against** using `terraform-json` if your
+application has any of the following requirements:
+
+* **Forwards-compatibility**: each version of this library represents a specific
+  snapshot of the [Terraform JSON output format](https://developer.hashicorp.com/terraform/internals/json-format),
+  and it often slightly lags Terraform itself. The library supports
```
Comment (radeksimko) [inline README.md:28]:
```suggestion
* **Forward-compatibility**: each version of this library represents a specific
  snapshot of the [Terraform JSON output format](https://developer.hashicorp.com/terraform/internals/json-format),
  and it often slightly lags behind Terraform itself. The library supports
```

**Review by radeksimko — APPROVED** [review]
